snapshotter/demux/cache/evict.go

```diff
@@ -0,0 +1,67 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"net"
+	"time"
+)
+
+// EvictionPolicy defines the interface for enforcing a cache eviction policy.
+type EvictionPolicy interface {
+	Enforce(key string)
+}
+
+type evictionPolicy struct {
+	evict chan string
+	stop  chan struct{}
+}
+
+// EvictOnConnectionFailurePolicy defines an eviction policy where entries are evicted
+// from cache after a failed connection attempt occurs.
+type EvictOnConnectionFailurePolicy struct {
+	evictionPolicy
+
+	dial      func(context.Context, string) (net.Conn, error)
+	frequency time.Duration
+}
+
+// NewEvictOnConnectionFailurePolicy creates a new instance with
+func NewEvictOnConnectionFailurePolicy(evictChan chan string, stopCondition chan struct{}, dial func(context.Context, string) (net.Conn, error), frequency time.Duration) EvictionPolicy {
+	return &EvictOnConnectionFailurePolicy{evictionPolicy: evictionPolicy{evict: evictChan, stop: stopCondition}, dial: dial, frequency: frequency}
+}
+
+// Enforce launches a go routine which periodically attempts to dial the cached entry
+// using the provided dial function.
+//
+// On connection failure, the entry will be evicted from cache.
+func (p *EvictOnConnectionFailurePolicy) Enforce(key string) {
+	go func(dial func(context.Context, string) (net.Conn, error)) {
+		ticker := time.NewTicker(p.frequency)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				if _, err := dial(context.Background(), key); err != nil {
```

Comment:
Is it okay to pass context.Background()? Can dial() take the context's cancellation signal?

Reply:
So if we had something like this where the ticker fired at a very frequent ratio, then cancel would stop all subsequent threads performing dials?

```
		ctx, cancel := context.WithCancel(context.Background())

		for {
			select {
			case <-ticker.C:
				conn, err := dial(ctx, key)
				if err != nil {
					p.evict <- key
					cancel()
					return
				}
				conn.Close()
			case <-p.stop:
				cancel()
				return
			}
		}
```

Reply:
We can't send `p.stop` because dial is the same function used by the cache to dial. I feel like the `RemoteSnapshotter` should be reword to use some future-style connection so we don't block on dialing, but that is a separate PR.

Reply:
Sorry, I meant, we know that this dial() would fail and this code is using the failure as a signal to remove dead VMs. If so, dial() should fail after a few seconds/minutes instead of waiting indefinitely.

Passing `context.Background()` *indicates* that this function doesn't have any deadlines and would't be cancelled.

Reply:
Hmm yes you are right. We have talked before about user configured timeouts. I think it makes sense to potentially try that here where the context has a timeout specified by customer. So the knobs would be

1. dial timeout (overall)
2. dial retry (should be much less than dial timeout)

Is it enough to infer individual dial attempt timeouts or would user specify that as well?

Reply:
Added a separate commit with dialer timeout threaded through to the eviction mechanism.